### PR TITLE
Keyword clustering

### PR DESCRIPTION
**Pull request by yechenh**

### CODE REVIEW

File: keyword_clustering.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import argparse
+import csv
+import os
+
+"""Cluster keywords based on the result of summarization."""
```
Comment (XiangZeng2020) [inline keyword_clustering.py:19]: Move file comment before the imports. example: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/layers/base.py

File: keyword_clustering.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    input_file_path = os.path.expanduser(args.input_file_path)
+    if not os.path.isfile(input_file_path):
+        raise ValueError("Cannot find the input file.")
+
+    shortened_keywords_list = []
+    total_keyword_counts = 0
+    
+    with open(input_file_path) as f:
+        read_tsv = csv.reader(f, delimiter="\t") 
+        model_name_list = next(read_tsv)[1:]
```
Comment (XiangZeng2020) [inline keyword_clustering.py:32]: reduce one indent level: 

with open(input_file_path) as f:
        read_tsv = csv.reader(f, delimiter="\t") 
model_name_list = next(read_tsv)[1:]
for i in range(len(model_name_list)):
      shortened_keywords_list.append({})
...

File: keyword_clustering.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    for index, model in enumerate(model_name_list):
+        clustered_keywords = 0
+        print(model)
+        if len(model_name_list) == 1:
+            model = ""
+        this_output_file_path = "/".join(output_file_path.split("/")[:-1]) + "/" + \
+            file_name + "_" + model + ".tsv"
+        with open(os.path.expanduser(this_output_file_path), 'wt') as f:
+            tsv_writer = csv.writer(f, delimiter='\t')
+            for shortened_keywords, keyword_list in shortened_keywords_list[index].items():
```
Comment (XiangZeng2020) [inline keyword_clustering.py:62]: reduce one indent level

File: keyword_clustering.py (hunk 1 above)
Comment (yechenh) [reply keyword_clustering.py]: Done.

File: keyword_clustering.py (hunk 2 above)
Comment (yechenh) [reply keyword_clustering.py]: Done.

File: keyword_clustering.py (hunk 3 above)
Comment (yechenh) [reply keyword_clustering.py]: Done.

File: keyword_clustering.py [hunk 4]
```diff
@@ -27,25 +27,23 @@ def main(args):
...
+        total_keyword_counts += 1
+        for index, shortened_keyword in enumerate(line[1:]):
+            shortened_keyword = shortened_keyword.lower()
+            if shortened_keyword == "":
+                continue
+            if shortened_keyword not in shortened_keywords_list[index]:
+                shortened_keywords_list[index][shortened_keyword] = [line[0]]
+            else:
+                shortened_keywords_list[index][shortened_keyword].append(line[0])
+    f.close()
```
Comment (XiangZeng2020) [inline keyword_clustering.py:46]: same here, the with statement is good, you can just keep it.

File: keyword_clustering.py [hunk 5]
```diff
@@ -57,14 +55,16 @@ def main(args):
...
+        
+        f = open(os.path.expanduser(this_output_file_path), 'wt')
+        tsv_writer = csv.writer(f, delimiter='\t')
+        for shortened_keywords, keyword_list in shortened_keywords_list[index].items():
+            if len(keyword_list) > 1:
+                clustered_keywords += len(keyword_list)        
+                keyword_list.insert(0, len(keyword_list))
+                keyword_list.insert(0, shortened_keywords)
+                tsv_writer.writerow(keyword_list)  
+        f.close()
```
Comment (XiangZeng2020) [inline keyword_clustering.py:67]: revert to the original version, sorry about the confusion.

File: keyword_clustering.py (hunk 4 above)
Comment (yechenh) [reply keyword_clustering.py]: Done

File: keyword_clustering.py (hunk 5 above)
Comment (yechenh) [reply keyword_clustering.py]: Done

File: keyword_clustering.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+            model = ""
+        this_output_file_path = "/".join(output_file_path.split("/")[:-1]) + "/" + \
+            file_name + "_" + model + ".tsv"
+        
+        with open(os.path.expanduser(this_output_file_path), 'wt') as f:
+            tsv_writer = csv.writer(f, delimiter='\t')
+            for shortened_keywords, keyword_list in shortened_keywords_list[index].items():
+                if len(keyword_list) > 1:
+                    clustered_keywords += len(keyword_list)        
+                    keyword_list.insert(0, len(keyword_list))
```
Comment (liangji-google) [inline keyword_clustering.py:63]: Define a new list and add elements from beginning.
From here https://wiki.python.org/moin/TimeComplexity
Internally, a list is represented as an array. Insert at position 0 takes lots of time to move every elements.

File: keyword_clustering.py [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+        print(clustered_keywords, "of", total_keyword_counts ,"keywords are clustered")    
+        
+
+if __name__ == "__main__":
+    """Cluster keywords based on their outputs in the LaserTagger model.
+    
+    usage: keyword_clustering.py [-h] input_file_path output_file_path
+
+    positional arguments:
+      input_file_path   the directory of the LaserTagger model output
```
Comment (liangji-google) [inline keyword_clustering.py:76]: Given some detail about input_file format, what is each columns means and give a small example of input_file.

Comment (yechenh) [reply keyword_clustering.py:75]: Done

File: keyword_clustering.py (hunk 6 above)
Comment (yechenh) [reply keyword_clustering.py]: Done

File: keyword_clustering.py [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+      output_file_path  Where the clustering result will be stored. If there is output from only one model 
+      in the input file, then the output will be savedat this path. Otherwise, the name of the model will be
```
[the remark this answers is not in the document]
Comment (yechenh) [reply keyword_clustering.py]: Done

**Pull request merged**